system/jlib/jsecrets.cpp

```diff
@@ -475,16 +538,28 @@ class SecretCache
         else
         {
             //Insert an entry with a null value that is marked as out of date
-            result = new SecretCacheEntry(now);
+            result = new SecretCacheEntry(now, secretKey.c_str());
             secrets.emplace(secretKey, result);
         }
         return result;
     }
 
-    void updateSecret(SecretCacheEntry * match, IPropertyTree * value, cache_timestamp now)
+
+    void gatherPendingRefresh(std::vector<SecretCacheEntry *> & pending, cache_timestamp when)
+    {
+        CriticalBlock block(cs);
+        for (auto & entry : secrets)
+        {
+            SecretCacheEntry * secret = entry.second.get();
+            if (secret->isActive() && secret->needsRefresh(when))
```

Comment:
This doesn't feel right - just because a secret has not been read recently doesn't mean that the next time it IS used it's ok to use an old version, does it?

Reply:
Next time it is actually used it will check to see if it needs a refresh, and if it does it will go and get the new value.  This PR doesn't change that behaviour, it preemptively refreshes secrets that are in active use behind the scenes to avoid roxie pausing when it needs to access them.